src/main/scala/io/iohk/ethereum/consensus/Ethash.scala

```diff
@@ -0,0 +1,234 @@
+package io.iohk.ethereum.consensus
+
+import java.math.BigInteger
+import java.nio.{ByteBuffer, ByteOrder}
+import java.util
+
+import akka.util.ByteString
+import io.iohk.ethereum.crypto.{kec256, kec512}
+import io.iohk.ethereum.domain.BlockHeader
+import io.iohk.ethereum.utils.ByteUtils
+import org.spongycastle.util.BigIntegers
+import org.spongycastle.util.encoders.Hex
+
+import scala.annotation.tailrec
+
+object Ethash {
+
+  // scalastyle:off magic.number
+
+  // value used in Fowler–Noll–Vo hash function
+  val FNV_PRIME: Int = 0x01000193
+
+  // bytes in word
+  val WORD_BYTES: Int = 4
+
+  // bytes in dataset at genesis
+  val DATASET_BYTES_INIT: Long = BigInt(2).pow(30).toLong
+
+  // dataset growth per epoch
+  val DATASET_BYTES_GROWTH: Long = BigInt(2).pow(23).toLong
+
+  // bytes in cache at genesis
+  val CACHE_BYTES_INIT: Long = BigInt(2).pow(24).toLong
+
+  // cache growth per epoch
+  val CACHE_BYTES_GROWTH: Long = BigInt(2).pow(17).toLong
+
+  // blocks per epoch
+  val EPOCH_LENGTH: Int = 30000
+
+  // width of mix
+  val MIX_BYTES: Int = 128
+
+  // hash length in bytes
+  val HASH_BYTES: Int = 64
+
+  // number of parents of each dataset element
+  val DATASET_PARENTS: Int = 256
+
+  // number of rounds in cache production
+  val CACHE_ROUNDS: Int = 3
+
+  // number of accesses in hashimoto loop
+  val ACCESSES: Int = 64
+
+  // scalastyle:on magic.number
+
+  def seed(blockNumber: Long): ByteString = {
+    (BigInt(0) until epoch(blockNumber))
+      .foldLeft(ByteString(Hex.decode("00" * 32))) { case (b, _) => kec256(b) }
+  }
+
+  def epoch(blockNumber: Long): Long = blockNumber / EPOCH_LENGTH
+
+  def cacheSize(blockNumber: Long): Long = {
+    val sz = (CACHE_BYTES_INIT + CACHE_BYTES_GROWTH * epoch(blockNumber)) - HASH_BYTES
+    highestPrimeBelow(sz, HASH_BYTES)
+  }
+
+  def dagSize(blockNumber: Long): Long = {
+    val sz = DATASET_BYTES_INIT + DATASET_BYTES_GROWTH * epoch(blockNumber) - MIX_BYTES
+    highestPrimeBelow(sz, MIX_BYTES)
+  }
+
+  private def highestPrimeBelow(n: Long, len: Long): Long = {
+    if (isPrime(n / len)) n
+    else highestPrimeBelow(n - 2 * len, len)
+  }
+
+  private def isPrime(n: BigInt): Boolean =
+    !((2 until n.toInt-1) exists (n % _ == 0))
```

Comment:
It seems that `isPrime(1)` returns true, could that bring any issues to the implementation?

Reply:
Also, a simple optimisation: it's enough to iterate through numbers up to (and including) `floor(sqrt(n))`.

It this function used often? If yes, and if we know that `n < 2 ^ 31`, then [Fermat primality test](https://en.wikipedia.org/wiki/Primality_test#Fermat_primality_test) would be a simple-to-code, yet very efficient implementation. A TODO comment is also fine :)

Reply:
I've changed the implementation a bit so that it works with BigInts